### PR TITLE
Filament_runout_sensor: Module to use one or mode filament sensors

### PR DESCRIPTION
**Pull request by paulodrugos**
Corrected pull request after I messed up the #1076 , sorry.

Signed-off-by: Paulo Drugos paulodrugos@gmail.com

### CODE REVIEW

**Comment by Arksine** [comment]
Hi Paul, it seems we have been working toward the same goal.  I am currently working on the PAT9125EL laser filament sensor module, and it doing so I created a generic 'filament_sensor' module that duplicates and extends the work you have done here.

[filament_sensor.py](https://github.com/PrusaOwners/klipper/blob/pat9125-testing/klippy/extras/filament_sensor.py)

The general idea is that it can support various underlying implementations, such as the endstop switch sensor or the Prusa PAT9125.  It could be extended to support other sensors such as the Duet PAT9125 or the rumored PAT9130 Prusa is working on now.  This module supports insert and runout detection, ie. it can optionally execute a gcode to load filament into an extruder when the callback has triggered.  It also exposes its functionality so internal modules, such as the MMU2, can be notified of a filament event.

Take a look and let me know what you think.

Eric

File: config/example-extras.cfg [hunk 1]
```diff
@@ -1094,3 +1094,13 @@
...
+
+
+# Filament run out sensor (one may define any number of sections with a
+# "filament_runout_sensor" prefix). When trigered, a message will be sent
+# back and the pause command will be executed in octoprint.
+# By default pin HIGH=FILAMENT PRESENT
+# You can check the state of the sensor using the command
+# QUERY_FILAMENT_SENSOR SENSOR=sensor_name
+#[filament_runout_sensor sensor_name]
+#pin = ^ar4
```
Comment (alecplumb) [inline config/example-extras.cfg]: The safer configuration here would be `pin = ^!ar4`, with the runout switch connected normally-open between ground and the ar4 pin. That way if there is any wiring fault the sensor will read that there is no filament present.

**Comment by paulodrugos** [comment]
@Arksine Great! Makes more sense to have all the filament sensors in one single module. I looked over there and the only thing that's missing its the command to query the sensors, all the rest looks like it should work the same way that this module work for a simple switch sensor.

File: config/example-extras.cfg (hunk 1 above)
Comment (paulodrugos) [reply config/example-extras.cfg]: You're correct, but this is in a example file, it should not have the pin parameter predetermined. I think that the correction I must do its to remove the `^ar4` from this line.

**Comment by LazaroFilm** [comment]
What’s the benefit of having the sensor managed by Klipper vs Octoprint via plugin?

**Comment by trevjonez** [comment]
My opinion as a random bystander: Smooth octoprint integration with klipper is a really good thing, and is unique to klipper as far as I have experienced. The cost of that integration now is that you can do a fresh klipper install without octoprint (or any desire to use it) and still end up with it front and center in your lcd menu (probably an indicator that the way menu's work needs some iteration). The last thing klipper needs IMO is something like octoprint knowing about or owning its sensor configurations. I would say that nearly any extras module being directly aware of octoprint is unacceptable in terms of software design.

**Comment by KevinOConnor** [comment]
FYI, I'm inclined to wait a few additional weeks to see how the additional support develops.  Let me know if there are concerns with this.

-Kevin

**Comment by paulodrugos** [comment]
Since @Arksine have duplicated my work into a more complete filament_sensor module, I think there's no point to continue with this one. Should I close the pull request?

**Comment by Arksine** [comment]
Just to add to this, I am in the process of testing the switch element now.  I have submitted a PR for a module that handles pause/resume functionality with the ability to capture and return to that position.  I don't think it will be long before I submit the filament sensor as well.

The PAT9125 will need a lot more testing before I'm comfortable submitting it, but it doesn't need to be submitted with the generic filament sensor and the switch implementation.

**Comment by alecplumb** [comment]
@Arksine Do you still plan to do a generic filament sensor using your pause_resume functionality?

**Comment by Arksine** [comment]
@alecplumb Yes.  Shouldn't be long.

**Comment by KevinOConnor** [comment]
FYI, PR #1301 was merged.

-Kevin

**Comment by ETE-Design** [comment]
Any news on this? And would it be possible to use multiple runout sensors?

**Pull request closed without merging**